File: Java/hydra/src/main/java/uk/co/shastra/hydra/messaging/listeners/StdListener.java
package uk.co.shastra.hydra.messaging.listeners;

import java.util.ArrayList;
import java.util.Date;
import rx.Observable;
import rx.Scheduler;
import rx.Subscription;
import rx.concurrency.NewThreadScheduler;
import rx.subjects.PublishSubject;
import rx.subjects.Subject;
import rx.util.functions.Action1;
import rx.util.functions.Func0;
import rx.util.functions.Func1;
import uk.co.shastra.hydra.messaging.TransportMessage;
import uk.co.shastra.hydra.messaging.messagefetchers.MessageFetcher;
import uk.co.shastra.hydra.messaging.messageids.MessageId;
import uk.co.shastra.hydra.messaging.messageids.MessageIdManager;
import uk.co.shastra.hydra.messaging.storage.ChangesWrapper;
import uk.co.shastra.hydra.messaging.storage.Provider;
import uk.co.shastra.hydra.messaging.storage.Store;
import uk.co.shastra.hydra.messaging.utils.Event;
import uk.co.shastra.hydra.messaging.utils.EventHandler;
import uk.co.shastra.hydra.messaging.utils.ListUtils;
import uk.co.shastra.hydra.messaging.utils.ObservableGenerator;

public class StdListener<TMessage extends TransportMessage> implements Listener<TMessage> {

    private static final long DefaultPollIntervalMs = 1000;
    private static final long DefaultBufferDelayMs = 0;
    
	private Provider provider;
	private MessageFetcher<TMessage> messageFetcher;
	private boolean closed = false;
	private Store store;
    private ArrayList<TMessage> messageBuffer = new ArrayList<TMessage>();
    private long lastSeq;
    private MessageId startId;
	private Subscription subscription;
    private Subject<TMessage, TMessage> subject = PublishSubject.create();
    private Scheduler scheduler;
	
	private long bufferDelayMs;
	private long pollIntervalMs;
	private MessageId lastId;

	
    public long getBufferDelayMs() { return bufferDelayMs; }
	public void setBufferDelayMs(long bufferDelayMs) { this.bufferDelayMs = bufferDelayMs; }
	public long getPollIntervalMs() { return pollIntervalMs; }
	public void setPollIntervalMs(long pollIntervalMs) { this.pollIntervalMs = pollIntervalMs; }
	/**
	 * @return The last Id raised to clients. While processing a message, this will be the Id of that message.
	 */
	public MessageId getLastId() { return lastId; }
	
	
	public StdListener(Provider provider, MessageFetcher<TMessage> messageFetcher)  { this(provider, messageFetcher, null, null, null); }
	public StdListener(Provider provider, MessageFetcher<TMessage> messageFetcher, MessageId startId) { this(provider, messageFetcher, startId, null, null); }
    public StdListener(Provider provider, MessageFetcher<TMessage> messageFetcher, MessageId startId, ListenerOptions listenerOptions) { this(provider, messageFetcher, startId, listenerOptions, null); }
	/**
     * Construct a Listener and start it polling
     * 
     * The polling interval is taken from Service.getConfig().getPollIntervalMs() and is dynamic: changes take effect after the next poll.
     * 
     * @param provider The Provider to use
     * @param messageFetcher MessageFetcher with which to poll
     * @param startId Only fetch messages with higher id than startId. Defaults to the id corresponding to now.
     * @param listenerOptions Default values for Listener options.
     * @param scheduler Scheduler to use for polling. Defaults to NewThreadScheduler.getInstance()
     */
    public StdListener(Provider provider, MessageFetcher<TMessage> messageFetcher, MessageId startId, ListenerOptions listenerOptions, Scheduler scheduler)
    {
        this.provider = provider;
        this.messageFetcher = messageFetcher;
        this.scheduler = scheduler == null ? NewThreadScheduler.getInstance() : scheduler;
        bufferDelayMs = listenerOptions == null ? DefaultBufferDelayMs : listenerOptions.bufferDelayMs;
        pollIntervalMs = listenerOptions == null || (listenerOptions.pollIntervalMs == null) ? DefaultPollIntervalMs : listenerOptions.pollIntervalMs;
        lastId = startId == null ? MessageIdManager.create(new Date(this.scheduler.now())) : startId;
        subscription = 
        	new ObservableGenerator<Iterable<TMessage>>(pollIntervalMs, pollIntervalMs, new Func0<Iterable<TMessage>>() {
				@Override public Iterable<TMessage> call() {
					return OnElapsed();
			}}, this.scheduler).
			getObservable().
			mapMany(new Func1<Iterable<TMessage>, Observable<TMessage>>() {
				// Flatten the list
				@Override public Observable<TMessage> call(Iterable<TMessage> messages) {
					return Observable.from(messages);
			}}).
			subscribe(new Action1<TMessage>() {
				@Override public void call(TMessage message) { onMessageInQueue(message); }
			});
    }
	
	private final Iterable<TMessage> noMessages = new ArrayList<TMessage>();

    private Iterable<TMessage> OnElapsed()
    {
        if (closed) return noMessages;

        try {
            return Poll();
        } catch (Exception e) {
            // TODO: detect what sort of error this was
            provider.serverError(store.getName());
            return noMessages;
        }
    }
    
	private Iterable<TMessage> Poll() {
        Store pollStore = provider.getStore(false);
        if (pollStore == null) {
            // Stores are all offline, or initialisation is incomplete. Do nothing and wait until a pollStore is available.
            return noMessages;
        }
        if (store == null || !pollStore.getName().equals(store.getName())) {
            // The server has changed, so reinitialise. As store is initially null, this will be also be called on the very first poll.
            store = pollStore;
            startId = getLastId();
            // Populate messageBuffer from startId
            lastSeq = store.getLastSeq();
            // Fetch messages from startId, ignoring ones after lastSeq.
            messageBuffer = ListUtils.IterableToArrayList(messageFetcher.messagesAfterIdUpToSeq(store, startId, lastSeq));
        } else {
            // Get changes from lastSeq. Fetch messages in the change set and put in messageBuffer.
            ChangesWrapper changes = store.getChanges(startId, lastSeq);
            lastSeq = changes.lastSeq;
            Iterable<MessageId> messageIds = changes.messageIds;
            if (messageIds.iterator().hasNext()) {
            	ArrayList<Iterable<TMessage>> argh = new ArrayList<Iterable<TMessage>>();
            	argh.add(messageBuffer);
            	argh.add(messageFetcher.messagesInSet(store, messageIds));
            	messageBuffer = ListUtils.Merge(argh);
            }
        }

        MessageId delayedId = MessageIdManager.create(new Date(scheduler.now() - getBufferDelayMs()));
        // Place messageBuffer messages <= delayedId in newMessages and remove them from messageBuffer
        // This slightly clunky structure prevents concurrent ArrayList access errors.
        ArrayList<TMessage> newMessages = new ArrayList<TMessage>();
        while (messageBuffer.size() > 0) {
        	TMessage message = messageBuffer.get(0);
        	if (message.getMessageId().compareTo(delayedId) <= 0) {
        		newMessages.add(message);
        		messageBuffer.remove(0);
        	} else {
        		break;
        	}
        }

        return newMessages;
	}
    
    protected void onMessageInQueue(TMessage message) {
        try {
            lastId = message.getMessageId();
            subject.onNext(message);
            messageInQueue.raise(this, message);
        } catch (Exception e) {
            // Swallow errors so that an error processing one message does not stop others from being processed.
            // TODO: log error
        }
	}
    
	@Override public Observable<TMessage> getObservable() { return subject; }
	
	private Event<TMessage> messageInQueue = new Event<TMessage>();
	@Override public void addMessageinQueueHandler(EventHandler<TMessage> handler) { messageInQueue.addHandler(handler); }
	@Override public void removeMessageinQueueHandler(EventHandler<TMessage> handler) { messageInQueue.removeHandler(handler); }
	
	@Override
	public void close() {
        if (!closed) {
	        subscription.unsubscribe();
	        subject.onCompleted();
	        closed = true;
        }
	}

}

File: DotNet/Messaging/Listeners/StdListener.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reactive.Concurrency;
using System.Reactive.Linq;
using System.Reactive.Subjects;
using Shastra.Hydra.Messaging.MessageFetchers;
using Shastra.Hydra.Messaging.MessageIds;
using Shastra.Hydra.Messaging.Storage;
using Shastra.Hydra.Messaging.Utils;

namespace Shastra.Hydra.Messaging.Listeners
{
    public class StdListener<TMessage> : IListener<TMessage> where TMessage : TransportMessage
    {
        private const long DefaultPollIntervalMs = 1000;
        private const long DefaultBufferDelayMs = 0;
        private readonly IProvider _provider;
        private readonly IMessageFetcher<TMessage> _messageFetcher;
        private readonly Subject<TMessage> _subject = new Subject<TMessage>();
        private List<TMessage> _messageBuffer = new List<TMessage>();
        private long _lastSeq;
        private IMessageId _startId;
        private IStore _store;
        private bool _disposed = false;
        private readonly ObservableGenerator<IEnumerable<TMessage>> _generator;
        private readonly IDisposable _messageSub;
        private readonly IScheduler _scheduler;

        public long BufferDelayMs { get; set; }
        public long PollIntervalMs { get; set; }

        /// <summary>
        /// The last Id raised to clients. While processing a message, this will be the Id of that message.
        /// </summary>
        public IMessageId LastId { get; private set; }

        /// <summary>
        /// Construct a Listener and start it polling.
        /// </summary>
        /// <param name="provider">The IProvider to use</param>
        /// <param name="messageFetcher">IMessageFetcher with which to poll.</param>
        /// <param name="startId">Only fetch messages with higher id than startId. Defaults to the id corresponding to now.</param>
        /// <param name="listenerOptions">Default values for Listener options.</param>
        /// <param name="scheduler">Scheduler to use for polling. Defaults to TaskPoolScheduler.Default.</param>
        /// <remarks>The polling interval is taken from Service.GetConfig().PollIntervalMs and is dynamic: changes take effect after the next poll.</remarks>
        public StdListener(IProvider provider, IMessageFetcher<TMessage> messageFetcher, IMessageId startId = null, ListenerOptions listenerOptions = null, IScheduler scheduler = null)
        {
            _provider = provider;
            _messageFetcher = messageFetcher;
            _scheduler = scheduler ?? TaskPoolScheduler.Default;
            BufferDelayMs = listenerOptions == null ? DefaultBufferDelayMs : listenerOptions.BufferDelayMs;
            PollIntervalMs = listenerOptions == null || !listenerOptions.PollIntervalMs.HasValue ? DefaultPollIntervalMs : listenerOptions.PollIntervalMs.Value;
            LastId = startId ?? MessageIdManager.Create(_scheduler.Now.UtcDateTime);
            // We'd like to do this, but it leaks memory.
            //_messageSub = Observable.Generate(true, _ => true, _ => false, _ => OnElapsed(), _ => TimeSpan.FromMilliseconds(PollIntervalMs), _scheduler).
            //              SelectMany(messages => messages).Subscribe(OnMessageInQueue);
            _generator = new ObservableGenerator<IEnumerable<TMessage>>(PollIntervalMs, OnElapsed, _scheduler);
            _messageSub = _generator.SelectMany(messages => messages).Subscribe(OnMessageInQueue);
        }

        #region Polling

        private readonly IEnumerable<TMessage> _noMessages = Enumerable.Empty<TMessage>();

        private IEnumerable<TMessage> OnElapsed()
        {
            if (_disposed) return _noMessages;

            try {
                return Poll();
            } catch (Exception) {
                // TODO: detect what sort of error this was
                _provider.ServerError(_store.Name);
                return _noMessages;
            }
        }

        private IEnumerable<TMessage> Poll()
        {
            var store = _provider.GetStore(false);
            if (store == null) {
                // Stores are all offline, or initialisation is incomplete. Do nothing and wait until a store is available.
                return _noMessages;
            }
            if (_store == null || store.Name != _store.Name) {
                // The server has changed, so reinitialise. As _store is initially null, this will be also be called on the very first poll.
                _store = store;
                _startId = LastId;
                // Populate _messageBuffer from _startId
                _lastSeq = _store.GetLastSeq();
                // Fetch messages from _startId, ignoring ones after _lastSeq.
                _messageBuffer = _messageFetcher.MessagesAfterIdUpToSeq(_store, _startId, _lastSeq).ToList();
            } else {
                // Get changes from _lastSeq. Fetch messages in the change set and put in _messageBuffer.
                var changes = _store.GetChanges(_startId, _lastSeq, out _lastSeq).ToList();
                if (changes.Any()) _messageBuffer = new List<IEnumerable<TMessage>> {_messageBuffer, _messageFetcher.MessagesInSet(_store, changes)}.Merge().ToList();
            }

            var delayedId = MessageIdManager.Create(_scheduler.Now.UtcDateTime.AddMilliseconds(-BufferDelayMs));
            var newMessages = _messageBuffer.TakeWhile(m => m.MessageId.CompareTo(delayedId) <= 0).ToList();
            _messageBuffer.RemoveRange(0, newMessages.Count);
            return newMessages;
        }

        #endregion

        #region MessageInQueue event

        public event Action<object, TMessage> MessageInQueue;

        protected virtual void OnMessageInQueue(TMessage message)
        {
            try {
                LastId = message.MessageId;
                _subject.OnNext(message);
                if (MessageInQueue != null) MessageInQueue(this, message);
            } catch (Exception) {
                // Swallow errors so that an error processing one message does not stop others from being processed.
                // TODO: log error
            }
        }

        #endregion

        # region Implementation of IObservable<TMessage>

        public IDisposable Subscribe(IObserver<TMessage> observer)
        {
            return _subject.Subscribe(observer);
        }

        #endregion

        #region Implementation of IDisposable

        // See http://msdn.microsoft.com/en-us/library/ms244737.aspx

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing)
        {
            if (disposing) {
                // free managed resources
                _messageSub.Dispose();
                _generator.Dispose();
                _subject.OnCompleted();
                _subject.Dispose();
                _disposed = true;
            }
            // free native resources if there are any.
        }

        #endregion
    }
}
